public/app/features/transformers/calculateHeatmap/heatmap.test.ts

```diff
@@ -121,4 +121,276 @@ describe('Heatmap transformer', () => {
+    });
+
+    it('converts ge labels to numeric buckets with linear scale', () => {
+      const frame = toDataFrame({
+        fields: [
+          { name: 'time', type: FieldType.time, values: [1000, 2000] },
+          {
+            name: '1',
+            type: FieldType.number,
+            labels: { ge: '1' },
+            values: [10, 15],
+          },
+          {
+            name: '10',
+            type: FieldType.number,
+            labels: { ge: '10' },
+            values: [20, 25],
+          },
+        ],
+      });
+
+      const heatmap = rowsToCellsHeatmap({
+        frame,
+        yBucketScale: { type: ScaleDistribution.Linear },
+        layout: HeatmapCellLayout.ge,
+      });
+
+      expect(heatmap.fields[1].values).toEqual([1, 10, 1, 10]);
+      expect(heatmap.fields[1].name).toBe('yMin'); // ge layout
+    });
+
+    it('generates yMax field for linear scale', () => {
+      const frame = toDataFrame({
+        fields: [
+          { name: 'time', type: FieldType.time, values: [1000] },
+          { name: '1', type: FieldType.number, values: [10] },
+          { name: '2', type: FieldType.number, values: [20] },
+          { name: '4', type: FieldType.number, values: [30] },
+        ],
+      });
+
+      const heatmap = rowsToCellsHeatmap({
+        frame,
+        yBucketScale: { type: ScaleDistribution.Linear },
+      });
+
+      // Should have yMin, yMax, and count fields
+      expect(heatmap.fields.length).toBe(4);
+      expect(heatmap.fields[2].name).toBe('yMax');
+      expect(heatmap.fields[2].type).toBe('number');
+
+      // yMax should be [2, 4, 8] (shifted buckets + calculated last bucket)
+      // Last bucket uses factor 2 (from 2→4) to estimate 4→8
+      expect(heatmap.fields[2].values).toEqual([2, 4, 8]);
+    });
+
+    it('clears yOrdinalDisplay for linear scale', () => {
+      const frame = toDataFrame({
+        fields: [
+          { name: 'time', type: FieldType.time, values: [1000] },
+          { name: '1', type: FieldType.number, values: [10] },
+          { name: '10', type: FieldType.number, values: [20] },
+        ],
+      });
+
+      const heatmap = rowsToCellsHeatmap({
+        frame,
+        yBucketScale: { type: ScaleDistribution.Linear },
+      });
+
+      expect(heatmap.meta?.custom?.yOrdinalDisplay).toBeUndefined();
+    });
+
+    it('preserves yOrdinalDisplay for non-linear scale (auto/ordinal)', () => {
+      const frame = toDataFrame({
+        fields: [
+          { name: 'time', type: FieldType.time, values: [1000] },
+          { name: 'low', type: FieldType.number, values: [10] },
+          { name: 'high', type: FieldType.number, values: [20] },
+        ],
+      });
+
+      const heatmap = rowsToCellsHeatmap({ frame });
+
+      expect(heatmap.meta?.custom?.yOrdinalDisplay).toEqual(['low', 'high']);
+    });
+
+    it('sets unit to undefined for linear scale (preserves original unit)', () => {
```

Comment:
should we test the "preserve" behavior too beyond the undefined case? like, set a unit on the bucket fields and check that they make it through?